### PR TITLE
Minimal libcloudproviders support

### PR DESCRIPTION
**Pull request by ckamm**
- Add it as an optional dependency for linux builds,
  controlled by WITH_LIBCLOUDPROVIDERS, which is enabled automatically
  if the dependencies are available.

- Add code to export sync folders through the library and to add a
  minimal "Settings..." menu action.

- Set up cloud provider registration by installing a file into a
  path like /usr/share/cloud-providers/.

- DBus name and path are derived from APPLICATION_REV_DOMAIN by default
  and may be overridden with APPLICATION_CLOUDPROVIDERS_DBUS_NAME and
  APPLICATION_CLOUDPROVIDERS_DBUS_PATH if necessary.

For  #7209

### CODE REVIEW

**Comment by ckamm** [comment]
@jnweiger maybe take a look at the linux-packaging related parts.

**Comment by guruz** [comment]
Since this is a minimal implementation, what about 2.6.1. ? :)

**Comment by ckamm** [comment]
@guruz I'm up for targeting 2.6.1, the main cost is setting up the build. Since it's a new feature I went with 2.7 by default.

**Comment by ckamm** [comment]
Notes on testing: I noticed my system gtk on ubuntu 18.04 does not use libcloudproviders. That's unfortunate and required a custom build of gtk with `--enable-cloudproviders`. I then run nautilus with `LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libcloudproviders.so.0.2.5 LD_LIBRARY_PATH=mygtk/gtk/.libs nautilus`.
- The ownCloud sync folders should show up in the location picker on the left
- The right-click menu of these entries should have a "Settings..." entry.

**Comment by guruz** [comment]
@ckamm You probably discussed this before but just wondering... loading with dlopen/QPlugin might have advantages for building (...but disavtanges for maintainability)?

File: src/gui/CMakeLists.txt [hunk 1]
```diff
@@ -189,6 +189,8 @@ IF( WIN32 )
     ENDIF()
 ENDIF()
 
+include(libcloudproviders/libcloudproviders.cmake)
```
Comment (dschmidt) [inline src/gui/CMakeLists.txt]: This is rather unusual.
*Usually* you would probably put it in CMakeLists.txt file, add it as a subdirectory and define a static target in there (with include dirs and linked libraries) against which you then can link the application. 
The way you did it works as well of course.

**Comment by ckamm** [comment]
Screenshot: 
image

**Comment by ckamm** [comment]
Unfortunate complication: libcloudproviders < 0.3.0 crashes when unexporting an account. I've not found a way to determine library version or work around it in code. **only enable if target has >=0.3.0**

File: src/gui/libcloudproviders/libcloudproviders.cmake [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+if(LIBCLOUDPROVIDERS_FOUND AND GIO_FOUND)
+    set(LIBCLOUDPROVIDERS_POSSIBLE "1")
+endif()
+
+# User visible config switch
+set(WITH_LIBCLOUDPROVIDERS ${LIBCLOUDPROVIDERS_POSSIBLE} CACHE BOOL "Whether to bulid with libcloudproviders")
+
+if(WITH_LIBCLOUDPROVIDERS AND NOT LIBCLOUDPROVIDERS_POSSIBLE)
+    message(FATAL_ERROR "Trying to enable libcloudproviders but dependencies are missing")
+endif()
```
[the remark this answers is not in the document]
Comment (ckamm) [reply src/gui/libcloudproviders/libcloudproviders.cmake]: No distro bug - it's just that we also need access to gio functions if we want to work with cloudproviders. I'll add a comment saying "to enable the cloudproviders feature we need the cloudproviders and gio-2.0 libraries"

[the remark this answers is not in the document]
Comment (dschmidt) [reply src/gui/libcloudproviders/libcloudproviders.cmake]: I find it odd to couple this so strongly here, I would do this way more naively: 
Search for both, enable feature if both are found. No need for microoptimizations.

[the remark this answers is not in the document]
Comment (dschmidt) [reply src/gui/libcloudproviders/libcloudproviders.cmake]: Ah, you use this to make it "switchable". 

Suggestion: use `OPTION()` for that.

File: src/gui/libcloudproviders/libcloudproviders.cpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        case SyncResult::SyncRunning:
+        case SyncResult::SyncAbortRequested:
+            status = CLOUD_PROVIDERS_ACCOUNT_STATUS_SYNCING;
+            break;
+        case SyncResult::Error:
+        case SyncResult::SetupError:
+            status = CLOUD_PROVIDERS_ACCOUNT_STATUS_ERROR;
+            break;
+        case SyncResult::Paused:
+            status = CLOUD_PROVIDERS_ACCOUNT_STATUS_IDLE; // questionable
```
Comment (ogoffart) [inline src/gui/libcloudproviders/libcloudproviders.cpp]: I guess we can't use `CLOUD_PROVIDERS_ACCOUNT_STATUS_INVALID`. Then the alternative would be `CLOUD_PROVIDERS_ACCOUNT_STATUS_ERROR`
But it is true nothing really fits.

File: src/gui/libcloudproviders/libcloudproviders.cpp [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            break;
+        }
+
+        // We don't do this in other status reporting, but it would make some sense?
+        //if (!account->_folder->accountState()->isConnected())
+        //    status = CLOUD_PROVIDERS_ACCOUNT_STATUS_ERROR;
```
Comment (ogoffart) [inline src/gui/libcloudproviders/libcloudproviders.cpp]: Same as with paused.
If `IDLE` means everything is fine and new files are going to be sync'ed,  then IDLE is not the right choice.
I'd say `ERROR` is better. even for paused, as it means something like:  some account are not synchronized.

Now, that depends on how these status are reported to the user.  I guess the actual information is in the message.

**Comment by guruz** [comment]
The conversation here mentions crashyness, is this relevant for us?
https://bugs.archlinux.org/task/62502

**Comment by ckamm** [comment]
@guruz I read this as "if you enable libcloudproviders support in gtk, some qt applications using GtkFileChooserDialog may crash" - which is probably a bug with how it's integrated into gtk, I doubt it applies to our use here.

**Comment by guruz** [comment]
@ckamm If you're saying that we won't cause this indirectly by depending on this library then great.
(And only gtk compiled with the library will cause the crash)

**Comment by guruz** [comment]
FYI @hefee @dragotin

**Pull request merged**